In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Observation:- UnicodeDecodeError occurs when Python tries to read a file (like a CSV) and can't figure out how to convert the bytes in the file into readable text (Unicode).

# This usually happens when:

# The file contains special or non-English characters (like ₹, é, ç, etc.)

In [2]:
df = pd.read_csv("C:\\Users\\sagar panda\\OneDrive\\Desktop\\uncleaned mobile dataset project.csv", encoding_errors='ignore')
df

Company Name            Model Name Mobile Weight   RAM     Front Camera  \
0          Apple       iPhone 16 128GB          174g   6GB             12MP   
1          Apple       iPhone 16 256GB          174g   6GB             12MP   
2          Apple       iPhone 16 512GB          174g   6GB             12MP   
3          Apple  iPhone 16 Plus 128GB          203g   6GB             12MP   
4          Apple  iPhone 16 Plus 256GB          203g   6GB             12MP   
..           ...                   ...           ...   ...              ...   
925         Poco          Pad 5G 128GB          571g   8GB              8MP   
926         Poco          Pad 5G 256GB          571g   8GB              8MP   
927      Samsung  Galaxy Z Fold6 256GB          239g  12GB  10MP, 4MP (UDC)   
928      Samsung  Galaxy Z Fold6 512GB          239g  12GB  10MP, 4MP (UDC)   
929      Samsung    Galaxy Z Fold6 1TB          239g  12GB  10MP, 4MP (UDC)   

    Back Camera            Processor Battery Capacity  Screen Size  \
0          48MP           A17 Bionic         3,600mAh   6.1 inches   
1          48MP           A17 Bionic         3,600mAh   6.1 inches   
2          48MP           A17 Bionic         3,600mAh   6.1 inches   
3          48MP           A17 Bionic         4,200mAh   6.7 inches   
4          48MP           A17 Bionic         4,200mAh   6.7 inches   
..          ...                  ...              ...          ...   
925         8MP  Snapdragon 7s Gen 2        10,000mAh  12.1 inches   
926         8MP  Snapdragon 7s Gen 2        10,000mAh  12.1 inches   
927        50MP   Snapdragon 8 Gen 3          4400mAh   7.6 inches   
928        50MP   Snapdragon 8 Gen 3          4400mAh   7.6 inches   
929        50MP   Snapdragon 8 Gen 3          4400mAh   7.6 inches   

    Launched Price (Pakistan) Launched Price (India) Launched Price (China)  \
0                 PKR 224,999             INR 79,999              CNY 5,799   
1                 PKR 234,999             INR 84,999              CNY 6,099   
2                 PKR 244,999             INR 89,999              CNY 6,499   
3                 PKR 249,999             INR 89,999              CNY 6,199   
4                 PKR 259,999             INR 94,999              CNY 6,499   
..                        ...                    ...                    ...   
925                PKR 66,220             INR 23,999              CNY 2,099   
926                PKR 71,220             INR 25,999              CNY 2,299   
927               PKR 604,999            INR 164,999                 13,999   
928               PKR 544,999            INR 176,999             CNY 15,999   
929             Not available            INR 200,999             CNY 17,999   

    Launched Price (USA) Launched Price (Dubai)  Launched Year  
0                USD 799              AED 2,799           2024  
1                USD 849              AED 2,999           2024  
2                USD 899              AED 3,199           2024  
3                USD 899              AED 3,199           2024  
4                USD 949              AED 3,399           2024  
..                   ...                    ...            ...  
925              USD 280              AED 1,029           2024  
926              USD 300              AED 1,099           2024  
927            USD 1,899              AED 7,199           2024  
928             USD 1719              AED 7,699           2024  
929            USD 2,259              AED 8,699           2024  

[930 rows x 15 columns]

In [3]:
df.shape

(930, 15)

# Observation:- Total no.of rows=930 and Total no.of columns=15

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 930 entries, 0 to 929
Data columns (total 15 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Company Name               930 non-null    object
 1   Model Name                 930 non-null    object
 2   Mobile Weight              930 non-null    object
 3   RAM                        930 non-null    object
 4   Front Camera               930 non-null    object
 5   Back Camera                930 non-null    object
 6   Processor                  930 non-null    object
 7   Battery Capacity           930 non-null    object
 8   Screen Size                930 non-null    object
 9   Launched Price (Pakistan)  930 non-null    object
 10  Launched Price (India)     930 non-null    object
 11  Launched Price (China)     930 non-null    object
 12  Launched Price (USA)       930 non-null    object
 13  Launched Price (Dubai)     930 non-null    object
 14  Launched Y

# Observation:- All column datatype is in string and 1 column(Launched Year) is in int data type

In [5]:
df.isnull().sum()

Company Name                 0
Model Name                   0
Mobile Weight                0
RAM                          0
Front Camera                 0
Back Camera                  0
Processor                    0
Battery Capacity             0
Screen Size                  0
Launched Price (Pakistan)    0
Launched Price (India)       0
Launched Price (China)       0
Launched Price (USA)         0
Launched Price (Dubai)       0
Launched Year                0
dtype: int64

# Observation:- No null values in any column.

In [6]:
df.columns.tolist()

['Company Name',
 'Model Name',
 'Mobile Weight',
 'RAM',
 'Front Camera',
 'Back Camera',
 'Processor',
 'Battery Capacity',
 'Screen Size',
 'Launched Price (Pakistan)',
 'Launched Price (India)',
 'Launched Price (China)',
 'Launched Price (USA)',
 'Launched Price (Dubai)',
 'Launched Year']

# Observation :- No need to use .strip() to remove whitespace in front and back of any column.

In [7]:
df["Company Name"].unique()

array(['Apple', 'Samsung', 'OnePlus', 'Vivo', 'iQOO', 'Oppo', 'Realme',
       'Xiaomi', 'Lenovo', 'Motorola', 'Huawei', 'Nokia', 'Sony',
       'Google', 'Tecno', 'Infinix', 'Honor', 'POCO', 'Poco'],
      dtype=object)

In [8]:
df["Company Name"].nunique()

19

# Observation:- Total 19 unique company name are there

In [9]:
df["Company Name"].value_counts()

Company Name
Oppo        129
Apple        97
Honor        91
Samsung      88
Vivo         86
Realme       69
Motorola     62
Infinix      56
OnePlus      53
Huawei       42
Tecno        39
POCO         30
Xiaomi       27
Google       21
Lenovo       15
Nokia        11
Sony          9
iQOO          3
Poco          2
Name: count, dtype: int64

# Observation:- Here "poco"(only 2 values) have spelling mistake it should be "POCO" and "POCO" having (30 values).

In [10]:
df[df["Company Name"]=="Poco"]

Company Name    Model Name Mobile Weight  RAM Front Camera Back Camera  \
925         Poco  Pad 5G 128GB          571g  8GB          8MP         8MP   
926         Poco  Pad 5G 256GB          571g  8GB          8MP         8MP   

               Processor Battery Capacity  Screen Size  \
925  Snapdragon 7s Gen 2        10,000mAh  12.1 inches   
926  Snapdragon 7s Gen 2        10,000mAh  12.1 inches   

    Launched Price (Pakistan) Launched Price (India) Launched Price (China)  \
925                PKR 66,220             INR 23,999              CNY 2,099   
926                PKR 71,220             INR 25,999              CNY 2,299   

    Launched Price (USA) Launched Price (Dubai)  Launched Year  
925              USD 280              AED 1,029           2024  
926              USD 300              AED 1,099           2024

# Observation:- To check the index number of company name having "Poco" value which is [[925,926]:0]

In [11]:
df.iloc[[925,926],0]=["POCO","POCO"]

# Observation:- Replace the Poco by POCO

In [12]:
df["Company Name"].mode()

0    Oppo
Name: Company Name, dtype: object

# Observation:- Maximum mobile are from oppo

In [13]:
df["Company Name"].describe()

count      930
unique      18
top       Oppo
freq       129
Name: Company Name, dtype: object

In [14]:
df["Model Name"].unique()

array(['iPhone 16 128GB', 'iPhone 16 256GB', 'iPhone 16 512GB',
       'iPhone 16 Plus 128GB', 'iPhone 16 Plus 256GB',
       'iPhone 16 Plus 512GB', 'iPhone 16 Pro 128GB',
       'iPhone 16 Pro 256GB', 'iPhone 16 Pro 512GB',
       'iPhone 16 Pro Max 128GB', 'iPhone 16 Pro Max 256GB',
       'iPhone 16 Pro Max 512GB', 'iPhone 15 128GB', 'iPhone 15 256GB',
       'iPhone 15 512GB', 'iPhone 15 Plus 128GB', 'iPhone 15 Plus 256GB',
       'iPhone 15 Plus 512GB', 'iPhone 15 Pro 128GB',
       'iPhone 15 Pro 256GB', 'iPhone 15 Pro 512GB',
       'iPhone 15 Pro Max 128GB', 'iPhone 15 Pro Max 256GB',
       'iPhone 15 Pro Max 512GB', 'iPhone 14 128GB', 'iPhone 14 256GB',
       'iPhone 14 512GB', 'iPhone 14 Plus 128GB', 'iPhone 14 Plus 256GB',
       'iPhone 14 Plus 512GB', 'iPhone 14 Pro 128GB',
       'iPhone 14 Pro 256GB', 'iPhone 14 Pro 512GB',
       'iPhone 14 Pro Max 128GB', 'iPhone 14 Pro Max 256GB',
       'iPhone 14 Pro Max 512GB', 'iPhone 13 mini 128GB',
       'iPhone 13 mini 256G

In [15]:
df["Model Name"].nunique()

908

In [16]:
df["Model Name"].duplicated().sum()

22

# Observation:- Out of 930, 908 are unique model name and 22 are duplicated.

In [17]:
df["Model Name"].value_counts()

Model Name
Pad 128GB             3
Pad 2 256GB           3
K7x 128GB             2
K10 5G 128GB          2
K9x 256GB             2
                     ..
V27 Pro 256GB         1
V25 Pro 128GB         1
V25 Pro 256GB         1
X90 Pro 256GB         1
Galaxy Z Fold6 1TB    1
Name: count, Length: 908, dtype: int64

In [18]:
df["Model Name"].mode()

0      Pad 128GB
1    Pad 2 256GB
Name: Model Name, dtype: object

# Observation:- These two models are more frequent
# 0      Pad 128GB
# 1    Pad 2 256GB

In [19]:
df.rename(columns={"Mobile Weight":"Mobile Weight(in gram)"},inplace=True)

# Observation:- Replace the column name from "Mobile Weight" to "Mobile Weight(in gram)".

In [20]:
df["Mobile Weight(in gram)"]=df["Mobile Weight(in gram)"].str.replace("g","")

In [21]:
df["Mobile Weight(in gram)"]=df["Mobile Weight(in gram)"].astype(float)

# Observation:- Inside this column 2 value were in decimal therefore we will have to convert its datatype into "float".

In [22]:
df[df["Mobile Weight(in gram)"]=="300.5"]

Empty DataFrame
Columns: [Company Name, Model Name, Mobile Weight(in gram), RAM, Front Camera, Back Camera, Processor, Battery Capacity, Screen Size, Launched Price (Pakistan), Launched Price (India), Launched Price (China), Launched Price (USA), Launched Price (Dubai), Launched Year]
Index: []

In [23]:
df["Mobile Weight(in gram)"].unique()

array([174. , 203. , 206. , 221. , 171. , 172. , 140. , 204. , 238. ,
       135. , 164. , 189. , 228. , 194. , 188. , 226. , 177. , 208. ,
       458. , 490. , 300.5, 468. , 682. , 708. , 234. , 196. , 168. ,
       195. , 167. , 254. , 187. , 263. , 199. , 190. , 202. , 198. ,
       192. , 235. , 191. , 181. , 178. , 175. , 165. , 143. , 229. ,
       732. , 586. , 498. , 523. , 726. , 567. , 503. , 480. , 366. ,
       508. , 433. , 674. , 426. , 360. , 205. , 179. , 200. , 183. ,
       185. , 180. , 173. , 184. , 170. , 162. , 210. , 155. , 215. ,
       550. , 610. , 223. , 150. , 146. , 158. , 163. , 153. , 176. ,
       201. , 193. , 159. , 156. , 182. , 535. , 510. , 500. , 560. ,
       520. , 540. , 555. , 239. , 186. , 533. , 218. , 207. , 233. ,
       222.8, 440. , 482. , 466. , 499. , 372. , 465. , 209. , 166. ,
       169. , 295. , 255. , 197. , 225. , 220. , 241. , 245. , 161. ,
       580. , 147. , 151. , 212. , 213. , 216. , 222. , 250. , 450. ,
       420. , 280. ,

In [24]:
df["Mobile Weight(in gram)"].nunique()

148

# Observation:- Total 148 unique weight in "mobile weight" column

In [25]:
df["Mobile Weight(in gram)"].value_counts()

Mobile Weight(in gram)
190.0    68
195.0    64
185.0    29
192.0    26
180.0    25
         ..
151.0     1
212.0     1
540.0     1
216.0     1
360.0     1
Name: count, Length: 148, dtype: int64

# Observation:- Total 68 mobile phones whose weight are 190 and 190 will be the mode also for "Mobile Weight(in gram)".

In [26]:
df["Mobile Weight(in gram)"].describe()

count    930.000000
mean     228.267097
std      105.432503
min      135.000000
25%      185.000000
50%      194.000000
75%      208.000000
max      732.000000
Name: Mobile Weight(in gram), dtype: float64

# Observation:-
# Most mobile devices in your dataset are relatively lightweight (between 135g and 208g).
# The mean is much higher than the median (228 > 194), suggesting a right-skewed distribution — this happens due to outliers on the heavier side.
# The max weight of 732g is unusually high — consider:
# Verifying if this is a tablet or a data entry error.

In [27]:
q1=df["Mobile Weight(in gram)"].quantile(0.25)
q3=df["Mobile Weight(in gram)"].quantile(0.75)
IQR=q3-q1
ll=q1-1.5*IQR
ul=q3+1.5*IQR
outliers=df["Mobile Weight(in gram)"][(df["Mobile Weight(in gram)"]<ll) | (df["Mobile Weight(in gram)"]>ul)]
print("Q1 = ",q1)
print("Q3 = ",q3)
print("IQR = ",IQR)
print("lower limit = ",ll)
print("upper limit = ",ul)
print("Outliers = ",outliers)
outliers.shape[0]

Q1 =  185.0
Q3 =  208.0
IQR =  23.0
lower limit =  150.5
upper limit =  242.5
Outliers =  36     140.0
37     140.0
38     140.0
48     135.0
49     135.0
       ...  
892    485.0
893    590.0
894    530.0
925    571.0
926    571.0
Name: Mobile Weight(in gram), Length: 125, dtype: float64


125

# Observation:
# Lower Limit=150
# Upper Limit=242
# Total Outliers=125/930  values

In [28]:
df.rename(columns={"RAM":"RAM(in GB)"},inplace=True)

# Observation:- Rename the column "RAM" into "RAM(in GB)".

In [29]:
df["RAM(in GB)"]=df["RAM(in GB)"].str.replace("GB","")

# Observation:-remove in values"GB" inside the column RAM(in GB)

In [30]:
df["RAM(in GB)"].value_counts()

RAM(in GB)
8         308
6         206
12        193
4         146
3          34
16         31
2           6
1.5         2
8 / 12      2
10          1
1           1
Name: count, dtype: int64

# Observation:- 
# 2. 8 / 12GB it is wrong value which is 2 times in this column.

In [31]:
df[df["RAM(in GB)"]=="8 / 12"]

Company Name Model Name  Mobile Weight(in gram) RAM(in GB) Front Camera  \
641       Huawei    P60 Pro                   200.0     8 / 12         13MP   
642       Huawei    P60 Art                   206.0     8 / 12         13MP   

                                           Back Camera  \
641  48MP (wide) + 13MP (ultrawide) + 48MP (telephoto)   
642  48MP (wide) + 40MP (ultrawide) + 48MP (telephoto)   

                  Processor Battery Capacity  Screen Size  \
641  Snapdragon 8+ Gen 1 4G         4,815mAh  6.67 inches   
642  Snapdragon 8+ Gen 1 4G         5,100mAh  6.73 inches   

    Launched Price (Pakistan) Launched Price (India) Launched Price (China)  \
641               PKR 189,999             INR 99,999              CNY 7,988   
642               PKR 219,999            INR 119,999              CNY 8,988   

    Launched Price (USA) Launched Price (Dubai)  Launched Year  
641            USD 1,099              AED 4,199           2023  
642            USD 1,299              AED 4,799           2023

# Observation:- To know the index number of wrong values in RAM(in GB) column.

In [32]:
df.drop([641,642],axis=0,inplace=True)

# Observation:- Drop the wrong values rows which is present in column "RAM(in GB)".

In [33]:
df["RAM(in GB)"].value_counts()

RAM(in GB)
8      308
6      206
12     193
4      146
3       34
16      31
2        6
1.5      2
10       1
1        1
Name: count, dtype: int64

In [34]:
df["RAM(in GB)"]=df["RAM(in GB)"].astype(float)

# Observation:- Convert the datatype of "RAM(in GB)" from string to float

In [35]:
df["RAM(in GB)"].unique()

array([ 6. ,  8. ,  4. ,  3. , 12. ,  2. ,  1.5, 16. , 10. ,  1. ])

In [36]:
df["RAM(in GB)"].nunique()

10

# Observation:- Total 10 different types of RAM inside column "RAM(in GB)"

In [37]:
df["RAM(in GB)"].describe()

count    928.000000
mean       7.784483
std        3.183086
min        1.000000
25%        6.000000
50%        8.000000
75%        8.000000
max       16.000000
Name: RAM(in GB), dtype: float64

#  Insights & Observations
# Most Common RAM Size:
# Since both the median and 75th percentile are 8 GB, it suggests 8 GB RAM is the most common configuration in this dataset.

# Skewness:

# The mean (7.78) is slightly less than the median (8), indicating a slight left-skew, though almost symmetrical.

# There are a few low-end phones (1 GB RAM) and some high-end ones (up to 16 GB), but most cluster around 8 GB.

# Market Segment Insight:

# Majority of devices are mid-range (6–8 GB RAM).

# A minority have very low or very high RAM, suggesting lower sales or availability of those models.

# Spread:

# The range is 15 GB (from 1 GB to 16 GB), but most devices fall within a narrower interquartile range (6–8 GB).

# The standard deviation (3.18) shows moderate variability, with a few extreme values pulling it slightly higher.



In [38]:
q1=df["RAM(in GB)"].quantile(0.25)
q3=df["RAM(in GB)"].quantile(0.75)
IQR=q3-q1
ll=q1-1.5*IQR
ul=q3+1.5*IQR
outliers=df["RAM(in GB)"][(df["RAM(in GB)"]<ll) | (df["RAM(in GB)"]>ul)]
print("Q1 = ",q1)
print("Q3 = ",q3)
print("IQR = ",IQR)
print("lower limit = ",ll)
print("upper limit = ",ul)
print("Outliers = ",outliers)
len(outliers)

Q1 =  6.0
Q3 =  8.0
IQR =  2.0
lower limit =  3.0
upper limit =  11.0
Outliers =  97     12.0
98     12.0
103    12.0
104    12.0
109    12.0
       ... 
890    12.0
893    12.0
927    12.0
928    12.0
929    12.0
Name: RAM(in GB), Length: 233, dtype: float64


233

# Observation:- Total outliers in RAM(in GB)=233

In [39]:
df["Front Camera"].unique()

array(['12MP', '12MP / 4K', '7MP', '10MP', '32MP', '13MP', '5MP', '16MP',
       '8MP', '12MP + 12MP', '2MP', '44MP', '24MP', '20MP+8MP', '20MP',
       '50MP', '25MP', '60MP', '10.7MP', 'Dual 32MP', 'Dual 60MP',
       '60MP (ultrawide) + 8MP (telephoto)', '60MP + 8MP', '11.1MP',
       '10.8MP', '10.5MP', '48MP', '42MP', '10MP, 4MP (UDC)'],
      dtype=object)

In [40]:
df["Front Camera"].nunique()

29

# Observation:- Total 29 types of front camera are present.

In [41]:
df["Front Camera"].value_counts()

Front Camera
16MP                                  211
32MP                                  207
8MP                                   165
12MP                                   81
5MP                                    47
13MP                                   41
12MP / 4K                              36
50MP                                   30
7MP                                    24
10MP                                   22
20MP                                   12
2MP                                    12
10.8MP                                  7
60MP                                    5
44MP                                    3
10.5MP                                  3
60MP + 8MP                              3
10MP, 4MP (UDC)                         3
10.7MP                                  2
60MP (ultrawide) + 8MP (telephoto)      2
25MP                                    2
12MP + 12MP                             2
20MP+8MP                                2
Dual 32MP            

# Observation:- Most cell phone has 16 & 32 MP front camera.

In [42]:
df["Back Camera"].unique()

array(['48MP', '50MP + 12MP', '48MP + 12MP', '12MP + 12MP', '12MP',
       '12MP + 12MP + 12MP', '8MP', '12MP + 10MP', '200MP + 12MP',
       '108MP + 12MP', '48MP + 8MP', '50MP + 5MP', '50MP + 2MP',
       '108MP + 8MP', '50MP + 8MP', '13MP + 2MP', '12MP + 16MP', '50MP',
       '16MP', '16MP + 5MP', '13MP', '13MP + 5MP', '16MP + 8MP',
       '13MP + 8MP', '13MP + 6MP', '5MP', '50MP + 48MP', '108MP',
       '64MP + 2MP', '48MP + 48MP', '48MP + 50MP', '48MP + 16MP',
       '48MP + 5MP', '64MP + 8MP', '20MP + 16MP', '16MP + 20MP',
       '50MP + 16MP', '200MP', '64MP', '13MP+2MP', '48MP + 64MP + 48MP',
       '50MP + 32MP + 48MP', '50MP + 50MP + 50MP', '50MP + 50MP',
       '50MP + 8MP + 2MP', '50MP + 50MP + 8MP', '50MP + 32MP + 8MP',
       '64MP + 8MP + 2MP', '8MP + 2MP', '50MP + 50MP + 64MP',
       '50MP + 50MP + 13MP', '50MP + 48MP + 32MP', '64MP + 32MP + 8MP',
       '50MP + 64MP + 8MP', '64MP + 8MP + 2MP + 2MP',
       '50MP + 13MP + 16MP + 2MP', '50MP + 16MP + 13MP + 2MP',
      

In [43]:
df["Back Camera"].nunique()

90

# Observation:- Total 90 types of Back Camera are present in phones.

In [44]:
df["Back Camera"].value_counts()

Back Camera
50MP                                                 182
13MP                                                  76
48MP                                                  60
64MP                                                  57
50MP + 2MP                                            39
                                                    ... 
48MP + 13MP + 8MP + 2MP                                1
16MP + 2MP + 2MP                                       1
12MP + 8MP + 2MP + 2MP                                 1
13MP+2MP                                               1
50MP (wide) + 40MP (ultrawide) + 48MP (telephoto)      1
Name: count, Length: 90, dtype: int64

# Observation:- 182 phones has 50MP back camera.

In [45]:
df["Back Camera"].describe()

count      928
unique      90
top       50MP
freq       182
Name: Back Camera, dtype: object

In [46]:
df["Processor"].unique().tolist()

['A17 Bionic',
 'A17 Pro',
 'A16 Bionic',
 'A15 Bionic',
 'A14 Bionic',
 'A13 Bionic',
 'A11 Bionic',
 'A12 Bionic',
 'A12Z Bionic',
 'Exynos 2400',
 'Snapdragon 8 Gen 2',
 'Exynos 2200',
 'Snapdragon 8 Gen 1',
 'Exynos 1380',
 'MediaTek Dimensity 1080',
 'MediaTek Helio G99',
 'Exynos 850',
 'Exynos 1280',
 'MediaTek Helio P35',
 'Exynos 990',
 'Exynos 9825',
 'Snapdragon 450',
 'Exynos 7870',
 'Snapdragon 425',
 'Exynos 7570',
 'Snapdragon 653',
 'Snapdragon 625',
 'Snapdragon 617',
 'Snapdragon 888',
 'Snapdragon 695',
 'Unisoc T618',
 'MediaTek Helio P22T',
 'Snapdragon 778G',
 'Exynos 9810',
 'Spreadtrum SC8830',
 'Qualcomm MSM8916',
 'Snapdragon 8 Gen 3',
 'MediaTek Dimensity 9000',
 'Snapdragon 782G',
 'MediaTek Dimensity 6020',
 'Snapdragon 8+ Gen 1',
 'MediaTek Dimensity 1300',
 'MediaTek Dimensity 1200-AI',
 'Snapdragon 480',
 'Qualcomm Snapdragon 460',
 'Snapdragon 865',
 'Snapdragon 870',
 'MediaTek Dimensity 900',
 'MediaTek Dimensity 1200',
 'Snapdragon 765G',
 'Snapdrago

In [47]:
df["Processor"].nunique()

217

# Observation:- Total 217 different types of processor are present.

In [48]:
df["Processor"].value_counts()

Processor
Snapdragon 8 Gen 2         30
MediaTek Dimensity 810     22
MediaTek Helio G99         21
Snapdragon 8 Gen 3         19
MediaTek Helio G85         19
                           ..
Snapdragon 439              1
Snapdragon 652              1
MediaTek Helio G35          1
MediaTek MT6592             1
MediaTek Dimensity 8400     1
Name: count, Length: 217, dtype: int64

In [49]:
df["Processor"].describe()

count                    928
unique                   217
top       Snapdragon 8 Gen 2
freq                      30
Name: Processor, dtype: object

# Observation:- Mostly used processor is "Snapdragon 8 Gen 2" which has frequency=30.

In [50]:
df["Battery Capacity"].unique().tolist()

['3,600mAh',
 '4,200mAh',
 '4,400mAh',
 '4,500mAh',
 '3,200mAh',
 '4,300mAh',
 '4,325mAh',
 '2,438mAh',
 '3,240mAh',
 '3,095mAh',
 '4,352mAh',
 '2,227mAh',
 '2,815mAh',
 '3,687mAh',
 '3,110mAh',
 '3,046mAh',
 '3,969mAh',
 '2,716mAh',
 '2,658mAh',
 '3,174mAh',
 '2,942mAh',
 '7,608mAh',
 '8,612mAh',
 '5,124mAh',
 '7,812mAh',
 '9,720mAh',
 '10,307mAh',
 '5000mAh',
 '4800mAh',
 '4000mAh',
 '4700mAh',
 '3900mAh',
 '4500mAh',
 '3800mAh',
 '4400mAh',
 '3700mAh',
 '6000mAh',
 '4300mAh',
 '3500mAh',
 '4050mAh',
 '3000mAh',
 '3600mAh',
 '3300mAh',
 '2600mAh',
 '11200mAh',
 '10090mAh',
 '8400mAh',
 '8000mAh',
 '7040mAh',
 '5100mAh',
 '5050mAh',
 '7600mAh',
 '4510mAh',
 '4115mAh',
 '4085mAh',
 '9510mAh',
 '11000mAh',
 '5700mAh',
 '4100mAh',
 '3315mAh',
 '3260mAh',
 '2300mAh',
 '3055mAh',
 '4030mAh',
 '2000mAh',
 '3200mAh',
 '4200mAh',
 '4450mAh',
 '4600mAh',
 '4830mAh',
 '4870mAh',
 '8040mAh',
 '4805mAh',
 '5800mAh',
 '5600mAh',
 '6400mAh',
 '8360mAh',
 '4520mAh',
 '5,000mAh',
 '4,600mAh',
 '4,700

In [51]:
df.rename(columns={"Battery Capacity":"Battery Capacity(in mAh)"},inplace=True)

# Observation:-renamed the column from "Battery Capacity" to "Battery Capacity(in mAh)".

In [52]:
df["Battery Capacity(in mAh)"]=df["Battery Capacity(in mAh)"].str.replace("mAh","")

# Observation:- Removed "mAh" from the values of column "Battery Capacity(in mAh)".

In [53]:
df["Battery Capacity(in mAh)"].nunique()

147

# Observation:- Total 147 unique battery capacity are present before remove comma from the values

In [54]:
df["Battery Capacity(in mAh)"]=df["Battery Capacity(in mAh)"].str.replace(",","")

# Observation:- Removed comma from the values.

In [55]:
df["Battery Capacity(in mAh)"].nunique()

122

# Observation:- Total 122 unique battery capacity are present after remove comma from the values.

In [56]:
df["Battery Capacity(in mAh)"]=df["Battery Capacity(in mAh)"].astype(int)

# Observation :- Converted the datatype from string to int.

In [57]:
df["Battery Capacity(in mAh)"].describe()

count      928.000000
mean      5026.311422
std       1356.989874
min       2000.000000
25%       4400.000000
50%       5000.000000
75%       5073.750000
max      11200.000000
Name: Battery Capacity(in mAh), dtype: float64

#  Key Observations
# Most Common Size:

# Median = 5000 mAh → Typical for most Android smartphones today.

# Low-End Devices:

# Devices with 2000–4000 mAh may be budget phones or older models.

# High-End / Special Devices:

# Max = 11200 mAh → These are outliers, possibly:

# Phones with very large batteries.

# Tablets or gaming phones.

# Data entry errors (if suspicious).

# Spread:

# Standard deviation of ~1357 mAh is relatively large.

# Suggests a wide battery size distribution, which might need segmentation.

In [58]:
q1=df["Battery Capacity(in mAh)"].quantile(0.25)
q3=df["Battery Capacity(in mAh)"].quantile(0.75)
IQR=q3-q1
ll=q1-1.5*IQR
ul=q3+1.5*IQR
outliers=df["Battery Capacity(in mAh)"][(df["Battery Capacity(in mAh)"]<ll) | (df["Battery Capacity(in mAh)"]>ul)]
print("Q1 = ",q1)
print("Q3 = ",q3)
print("IQR = ",IQR)
print("lower limit = ",ll)
print("upper limit = ",ul)
print("Outliers = ",outliers)
len(outliers)

Q1 =  4400.0
Q3 =  5073.75
IQR =  673.75
lower limit =  3389.375
upper limit =  6084.375
Outliers =  12      3200
13      3200
14      3200
24      3200
25      3200
       ...  
892     9000
893    10500
894     8500
925    10000
926    10000
Name: Battery Capacity(in mAh), Length: 154, dtype: int32


154

# Observation:- Total outliers are 154.

In [59]:
df["Screen Size"].unique().tolist()

['6.1 inches',
 '6.7 inches',
 '5.4 inches',
 '5.8 inches',
 '6.5 inches',
 '10.9 inches',
 '10.2 inches',
 '7.9 inches',
 '11 inches',
 '12.9 inches',
 '13 inches',
 '6.8 inches',
 '6.6 inches',
 '7.6 inches',
 '6.4 inches',
 '6.9 inches',
 '6.3 inches',
 '5.3 inches',
 '6.0 inches',
 '5.5 inches',
 '5.7 inches',
 '5.2 inches',
 '14.6 inches',
 '12.4 inches',
 '8.7 inches',
 '10.5 inches',
 '8 inches',
 '10.1 inches',
 '6.74 inches',
 '6.72 inches',
 '7.8 inches',
 '6.55 inches',
 '6.43 inches',
 '6.49 inches',
 '6.52 inches',
 '6.78 inches',
 '6.59 inches',
 '6.44 inches',
 '6.41 inches',
 '6.01 inches',
 '6.67 inches',
 '6.28 inches',
 '11.61 inches',
 '6.31 inches',
 '6.58 inches',
 '6.38 inches',
 '6.56 inches',
 '5.88 inches',
 '6.22 inches',
 '5.0 inches',
 '6.51 inches',
 '6.35 inches',
 '6.53 inches',
 '6.39 inches',
 '6.47 inches',
 '10.4 inches',
 '12.3 inches',
 '7.82 inches',
 '6.83 inches',
 '11.6 inches',
 '12.1 inches',
 '6.82 inches',
 '7.1 inches',
 '11.5 inches',
 '6

In [60]:
df[df["Screen Size"]=='6.7 inches (main), 2.7 inches (external)']

Company Name  Model Name  Mobile Weight(in gram)  RAM(in GB) Front Camera  \
568     Motorola  Razr 128GB                   192.0         8.0         32MP   
569     Motorola  Razr 256GB                   192.0        12.0         32MP   

    Back Camera            Processor  Battery Capacity(in mAh)  \
568        50MP  Snapdragon 8+ Gen 1                      3500   
569        50MP  Snapdragon 8+ Gen 1                      3500   

                                  Screen Size Launched Price (Pakistan)  \
568  6.7 inches (main), 2.7 inches (external)               PKR 199,999   
569  6.7 inches (main), 2.7 inches (external)               PKR 219,999   

    Launched Price (India) Launched Price (China) Launched Price (USA)  \
568             INR 59,999              CNY 5,999              USD 999   
569             INR 64,999              CNY 6,499            USD 1,099   

    Launched Price (Dubai)  Launched Year  
568              AED 3,699           2023  
569              AED 3,999           2023

In [61]:
df["Screen Size"].value_counts()

Screen Size
6.7 inches               124
6.5 inches                76
6.67 inches               68
6.6 inches                64
6.78 inches               61
                        ... 
8.0 inches (unfolded)      1
7.8 inches (unfolded)      1
5.0 inches                 1
7.93 inches                1
6.79 inches                1
Name: count, Length: 97, dtype: int64

In [62]:
df["Launched Price (Pakistan)"].unique().tolist()

['PKR 224,999',
 'PKR 234,999',
 'PKR 244,999',
 'PKR 249,999',
 'PKR 259,999',
 'PKR 274,999',
 'PKR 284,999',
 'PKR 294,999',
 'PKR 314,999',
 'PKR 324,999',
 'PKR 344,999',
 'PKR 204,999',
 'PKR 214,999',
 'PKR 264,999',
 'PKR 304,999',
 'PKR 184,999',
 'PKR 194,999',
 'PKR 364,999',
 'PKR 174,999',
 'PKR 354,999',
 'PKR 384,999',
 'PKR 334,999',
 'PKR 159,999',
 'PKR 169,999',
 'PKR 179,999',
 'PKR 219,999',
 'PKR 239,999',
 'PKR 269,999',
 'PKR 289,999',
 'PKR 309,999',
 'PKR 164,999',
 'PKR 149,999',
 'PKR 79,999',
 'PKR 89,999',
 'PKR 49,999',
 'PKR 59,999',
 'PKR 69,999',
 'PKR 189,999',
 'PKR 279,999',
 'PKR 359,999',
 'PKR 399,999',
 'PKR 450,000',
 'PKR 480,000',
 'PKR 400,000',
 'PKR 430,000',
 'PKR 350,000',
 'PKR 380,000',
 'PKR 460,000',
 'PKR 360,000',
 'PKR 390,000',
 'PKR 320,000',
 'PKR 420,000',
 'PKR 300,000',
 'PKR 330,000',
 'PKR 500,000',
 'PKR 550,000',
 'PKR 530,000',
 'PKR 85,000',
 'PKR 95,000',
 'PKR 75,000',
 'PKR 70,000',
 'PKR 80,000',
 'PKR 60,000',
 'P

In [63]:
df["Launched Price (Pakistan)"].nunique()

157

# Observation:- to get 'Not available' value inside the column.

In [64]:
df[df["Launched Price (Pakistan)"]=='Not available']

Company Name          Model Name  Mobile Weight(in gram)  RAM(in GB)  \
929      Samsung  Galaxy Z Fold6 1TB                   239.0        12.0   

        Front Camera Back Camera           Processor  \
929  10MP, 4MP (UDC)        50MP  Snapdragon 8 Gen 3   

     Battery Capacity(in mAh) Screen Size Launched Price (Pakistan)  \
929                      4400  7.6 inches             Not available   

    Launched Price (India) Launched Price (China) Launched Price (USA)  \
929            INR 200,999             CNY 17,999            USD 2,259   

    Launched Price (Dubai)  Launched Year  
929              AED 8,699           2024

# Observation:- index number of not available value =929

In [65]:
df.rename(columns={"Launched Price (Pakistan)":"Launched Price(Pakistan) in PKR"},inplace=True)

# Observation:- Renamed the column from "Launched Price (Pakistan)" to "Launched Price(Pakistan) in PKR".

In [66]:
df["Launched Price(Pakistan) in PKR"]=df["Launched Price(Pakistan) in PKR"].str.replace("PKR","")

# Observation:- Removed the value "PKR" from column "Launched Price(Pakistan) in PKR".

In [67]:
df["Launched Price(Pakistan) in PKR"]=df["Launched Price(Pakistan) in PKR"].str.replace("Not available","0")

# Observation:- Replaced the value "Not available" by "0" from column "Launched Price(Pakistan) in PKR"

In [68]:
df["Launched Price(Pakistan) in PKR"]=df["Launched Price(Pakistan) in PKR"].str.replace(",","")

# Observation:- Removed ","  from column "Launched Price(Pakistan) in PKR"

In [69]:
df["Launched Price(Pakistan) in PKR"]=df["Launched Price(Pakistan) in PKR"].astype(int)

# Observation:- Convert the datatype from string to int.

In [70]:
df["Launched Price(Pakistan) in PKR"].describe()

count       928.000000
mean     125129.423491
std      101661.762836
min           0.000000
25%       54999.000000
50%       84999.000000
75%      179999.000000
max      604999.000000
Name: Launched Price(Pakistan) in PKR, dtype: float64

# Observation:- mean>median(right skewed data
# Standard Deviation: PKR 101,579
# Prices are highly varied, indicating presence of low, mid, and premium segments

In [71]:
df["Launched Price(Pakistan) in PKR"]=df["Launched Price(Pakistan) in PKR"].replace(0,df["Launched Price(Pakistan) in PKR"].median())

# Observation:- Replaced the value "0" by "median of that column" from column "Launched Price(Pakistan) in PKR"

In [72]:
df["Launched Price(Pakistan) in PKR"].nunique()

156

# Observation:- Total 157 unique price in df["Launched Price(Pakistan) in PKR"]

In [73]:
q1=df["Launched Price(Pakistan) in PKR"].quantile(0.25)
q3=df["Launched Price(Pakistan) in PKR"].quantile(0.75)
IQR=q3-q1
ll=q1-1.5*IQR
ul=q3+1.5*IQR
outliers=df["Launched Price(Pakistan) in PKR"][(df["Launched Price(Pakistan) in PKR"]<ll) | (df["Launched Price(Pakistan) in PKR"]>ul)]
print("Q1 = ",q1)
print("Q3 = ",q3)
print("IQR = ",IQR)
print("lower limit = ",ll)
print("upper limit = ",ul)
print("Outliers = ",outliers)
len(outliers)

Q1 =  54999.0
Q3 =  179999.0
IQR =  125000.0
lower limit =  -132501.0
upper limit =  367499.0
Outliers =  47     384999
59     384999
96     399999
97     450000
98     480000
99     400000
100    430000
102    380000
103    430000
104    460000
106    390000
109    420000
110    450000
112    380000
115    500000
116    550000
118    380000
119    480000
120    530000
661    389999
676    429999
679    449999
682    469999
708    399999
719    399999
847    369999
858    389999
869    389999
927    604999
928    544999
Name: Launched Price(Pakistan) in PKR, dtype: int32


30

# Observation:- Total 30 outlier in df["Launched Price(Pakistan) in PKR"]

In [74]:
df["Launched Price (India)"].unique().tolist()

['INR 79,999',
 'INR 84,999',
 'INR 89,999',
 'INR 94,999',
 'INR 104,999',
 'INR 99,999',
 'INR 114,999',
 'INR 109,999',
 'INR 124,999',
 'INR 74,999',
 'INR 119,999',
 'INR 134,999',
 'INR 69,999',
 'INR 144,999',
 'INR 69,900',
 'INR 74,900',
 'INR 84,900',
 'INR 94,900',
 'INR 119,900',
 'INR 129,900',
 'INR 139,900',
 'INR 149,900',
 'INR 64,900',
 'INR 99,900',
 'INR 109,900',
 'INR 89,900',
 'INR 54,900',
 'INR 29,900',
 'INR 39,900',
 'INR 49,900',
 'INR 71,900',
 'INR 159,900',
 'INR 179,900',
 'INR 199,900',
 'INR 104,900',
 'INR 114,900',
 'INR 79,900',
 'INR 154,900',
 'INR 174,900',
 'INR 134,900',
 'INR 27,999',
 'INR 32,999',
 'INR 22,999',
 'INR 19,999',
 'INR 24,999',
 'INR 18,499',
 'INR 21,999',
 'INR 14,999',
 'INR 23,999',
 'INR 26,999',
 'INR 17,999',
 'INR 20,999',
 'INR 12,999',
 'INR 15,999',
 'INR 18,999',
 'INR 1,04,999',
 'INR 77,999',
 'INR 64,999',
 'INR 59,999',
 'INR 28,999',
 'INR 17,499',
 'INR 18,990',
 'INR 9,990',
 'INR 32,990',
 'INR 22,990',
 'IN

In [75]:
df["Launched Price (India)"].nunique()

147

# Observation:- Total 147 unique values are presemt in column "Launched Price (India)"

In [76]:
df.rename(columns={"Launched Price (India)":"Launched Price(India) in INR"},inplace=True)

# Observation:- Renamed the column from "Launched Price (India)" to "Launched Price(India) in INR".

In [77]:
df["Launched Price(India) in INR"]=df["Launched Price(India) in INR"].str.replace("INR","")

# Observation:- Removed the value "INR" from column "Launched Price(India) in INR".

In [78]:
df["Launched Price(India) in INR"]=df["Launched Price(India) in INR"].str.replace(",","")

# Observation:- Removed "," from column "Launched Price(India) in INR"

In [79]:
df["Launched Price(India) in INR"]=df["Launched Price(India) in INR"].astype(int)

# Observation:- Convert the datatype from string to int.

In [80]:
df["Launched Price(India) in INR"].describe()

count       928.000000
mean      50448.753233
std       40908.504288
min        5999.000000
25%       19999.000000
50%       34999.000000
75%       74900.000000
max      274999.000000
Name: Launched Price(India) in INR, dtype: float64

# Observation:-
#  Mean > Median → Right-skewed: A few expensive phones inflate the average
# Interquartile Range (IQR): ₹54,901 → wide spread among mid-range phones
# Standard Deviation: ₹40,909 → Prices are highly dispersed, confirming market diversity

In [81]:
q1=df["Launched Price(India) in INR"].quantile(0.25)
q3=df["Launched Price(India) in INR"].quantile(0.75)
IQR=q3-q1
ll=q1-1.5*IQR
ul=q3+1.5*IQR
outliers=df["Launched Price(India) in INR"][(df["Launched Price(India) in INR"]<ll) | (df["Launched Price(India) in INR"]>ul)]
print("Q1 = ",q1)
print("Q3 = ",q3)
print("IQR = ",IQR)
print("lower limit = ",ll)
print("upper limit = ",ul)
print("Outliers = ",outliers)
len(outliers)

Q1 =  19999.0
Q3 =  74900.0
IQR =  54901.0
lower limit =  -62352.5
upper limit =  157251.5
Outliers =  94     159900
95     179900
96     199900
116    174900
324    159999
630    249999
634    199999
643    249999
657    249999
660    259999
661    274999
708    179999
719    169999
858    159999
869    159999
927    164999
928    176999
929    200999
Name: Launched Price(India) in INR, dtype: int32


18

# Observation:- Total 18 outlier in "Launched Price(India) in INR"

In [82]:
df.head(1)

Company Name       Model Name  Mobile Weight(in gram)  RAM(in GB)  \
0        Apple  iPhone 16 128GB                   174.0         6.0   

  Front Camera Back Camera   Processor  Battery Capacity(in mAh) Screen Size  \
0         12MP        48MP  A17 Bionic                      3600  6.1 inches   

   Launched Price(Pakistan) in PKR  Launched Price(India) in INR  \
0                           224999                         79999   

  Launched Price (China) Launched Price (USA) Launched Price (Dubai)  \
0              CNY 5,799              USD 799              AED 2,799   

   Launched Year  
0           2024

In [83]:
df["Launched Price (China)"].unique().tolist()

['CNY 5,799',
 'CNY 6,099',
 'CNY 6,499',
 'CNY 6,199',
 'CNY 6,999',
 'CNY 7,099',
 'CNY 7,499',
 'CNY 7,799',
 'CNY 8,199',
 'CNY 5,299',
 'CNY 5,599',
 'CNY 5,999',
 'CNY 5,699',
 'CNY 6,799',
 'CNY 7,299',
 'CNY 7,999',
 'CNY 8,699',
 'CNY 5,199',
 'CNY 5,499',
 'CNY 5,899',
 'CNY 6,299',
 'CNY 7,199',
 'CNY 8,099',
 'CNY 8,499',
 'CNY 8,999',
 'CNY 7,699',
 'CNY 8,299',
 'CNY 8,799',
 'CNY 9,199',
 'CNY 8,399',
 'CNY 7,399',
 'CNY 8,388',
 'CNY 8,788',
 'CNY 9,188',
 'CNY 9,788',
 'CNY 10,388',
 'CNY 10,088',
 'CNY 10,688',
 'CNY 11,288',
 'CNY 3,299',
 'CNY 3,699',
 'CNY 2,299',
 'CNY 2,699',
 'CNY 3,499',
 'CNY 3,799',
 'CNY 6,399',
 'CNY 9,799',
 'CNY 10,399',
 'CNY 11,199',
 'CNY 6,599',
 'CNY 11,999',
 'CNY 12,999',
 'CNY 10,999',
 'CNY 2,199',
 'CNY 2,499',
 'CNY 1,799',
 'CNY 1,999',
 'CNY 1,599',
 'CNY 1,499',
 'CNY 1,699',
 'CNY 1,199',
 'CNY 1,399',
 'CNY 2,099',
 'CNY 1,099',
 'CNY 1,299',
 'CNY 2,399',
 'CNY 999',
 'CNY 3,099',
 'CNY 1,899',
 'CNY 9,999',
 'CNY 9,499',

In [84]:
df["Launched Price (China)"].nunique()

136

# Observation:- Total 136 unique values are presemt in column "Launched Price (China)"

In [85]:
df.rename(columns={"Launched Price (China)":"Launched Price(China) in CNY"},inplace=True)

# Observation:- Renamed the column from "Launched Price (China)" to "Launched Price(China) in CNY".

In [86]:
df["Launched Price(China) in CNY"]=df["Launched Price(China) in CNY"].str.replace("CNY","")

# Observation:- Removed the value "CNY" from column "Launched Price(India) in CNY".

In [87]:
df["Launched Price(China) in CNY"]=df["Launched Price(China) in CNY"].str.replace(",","")

# Observation:- Removed "," from column "Launched Price(China) in CNY"

In [88]:
df["Launched Price(China) in CNY"]=df["Launched Price(China) in CNY"].astype(int)

# Observation:- Convert the datatype from string to int.

In [89]:
df["Launched Price(China) in CNY"].describe()

count      928.000000
mean      3811.372845
std       2760.446211
min        499.000000
25%       1699.000000
50%       2800.000000
75%       5499.000000
max      17999.000000
Name: Launched Price(China) in CNY, dtype: float64

# Observation:-
# Mean > Median :Right-skewed distribution, pulled up by high-end phones
# Interquartile Range (IQR): 800 indicates significant spread among typical devices
# Standard Deviation: 760 Moderate to high variation in prices

In [90]:
q1=df["Launched Price(China) in CNY"].quantile(0.25)
q3=df["Launched Price(China) in CNY"].quantile(0.75)
IQR=q3-q1
ll=q1-1.5*IQR
ul=q3+1.5*IQR
outliers=df["Launched Price(China) in CNY"][(df["Launched Price(China) in CNY"]<ll) | (df["Launched Price(China) in CNY"]>ul)]
print("Q1 = ",q1)
print("Q3 = ",q3)
print("IQR = ",IQR)
print("lower limit = ",ll)
print("upper limit = ",ul)
print("Outliers = ",outliers)
len(outliers)

Q1 =  1699.0
Q3 =  5499.0
IQR =  3800.0
lower limit =  -4001.0
upper limit =  11199.0
Outliers =  76     11288
115    11999
116    12999
120    11999
630    17999
634    14999
643    13999
657    12999
660    13499
661    14499
708    13999
719    12999
927    13999
928    15999
929    17999
Name: Launched Price(China) in CNY, dtype: int32


15

# Observation:- Total 15 outlier in "Launched Price(China) in CNY"

In [91]:
df.head(1)

Company Name       Model Name  Mobile Weight(in gram)  RAM(in GB)  \
0        Apple  iPhone 16 128GB                   174.0         6.0   

  Front Camera Back Camera   Processor  Battery Capacity(in mAh) Screen Size  \
0         12MP        48MP  A17 Bionic                      3600  6.1 inches   

   Launched Price(Pakistan) in PKR  Launched Price(India) in INR  \
0                           224999                         79999   

   Launched Price(China) in CNY Launched Price (USA) Launched Price (Dubai)  \
0                          5799              USD 799              AED 2,799   

   Launched Year  
0           2024

In [92]:
df["Launched Price (USA)"].unique().tolist()

['USD 799',
 'USD 849',
 'USD 899',
 'USD 949',
 'USD 999',
 'USD 1,049',
 'USD 1,099',
 'USD 1,199',
 'USD 1,299',
 'USD 1,399',
 'USD 699',
 'USD 1,249',
 'USD 749',
 'USD 599',
 'USD 329',
 'USD 429',
 'USD 399',
 'USD 499',
 'USD 1,599',
 'USD 1,799',
 'USD 1,899',
 'USD 449',
 'USD 349',
 'USD 299',
 'USD 249',
 'USD 199',
 'USD 179',
 'USD 169',
 'USD 129',
 'USD 1,499',
 'USD 549',
 'USD 229',
 'USD 149',
 'USD 649',
 'USD 99',
 'USD 219',
 'USD 139',
 'USD 159',
 'USD 269',
 'USD 319',
 'USD 239',
 'USD 279',
 'USD 189',
 'USD 259',
 'USD 470',
 'USD 500',
 'USD 380',
 'USD 400',
 'USD 420',
 'USD 360',
 'USD 340',
 'USD 320',
 'USD 270',
 'USD 290',
 'USD 250',
 'USD 220',
 'USD 240',
 'USD 300',
 'USD 200',
 'USD 260',
 'USD 180',
 'USD 160',
 'USD 210',
 'USD 230',
 'USD 130',
 'USD 170',
 'USD 190',
 'USD 150',
 'USD 440',
 'USD 330',
 'USD 350',
 'USD 280',
 'USD 310',
 'USD 634.99',
 'USD 790.77',
 'USD 374.90',
 'USD 399.00',
 'USD 429.00',
 'USD 349.00',
 'USD 379.00',


In [93]:
df["Launched Price (USA)"].nunique()

107

# Observation:- Total 107 unique values are presemt in column "Launched Price (USA)"

In [94]:
df.rename(columns={"Launched Price (USA)":"Launched Price(USA) in USD"},inplace=True)

# Observation:- Renamed the column from "Launched Price (USA)" to "Launched Price(USA) in USD".

In [95]:
df["Launched Price(USA) in USD"]=df["Launched Price(USA) in USD"].str.replace("USD","")

# Observation:- Removed the value "USD" from column "Launched Price(USA) in USD".

In [96]:
df["Launched Price(USA) in USD"]=df["Launched Price(USA) in USD"].str.replace(",","")

# Observation:- Removed "," from column "Launched Price(USA) in USD"

In [97]:
df["Launched Price(USA) in USD"]=df["Launched Price(USA) in USD"].astype(float)

# Observation:- Convert the datatype from string to int.

In [98]:
df["Launched Price(USA) in USD"].describe()

count      928.000000
mean       624.279806
std       1348.742522
min         79.000000
25%        250.000000
50%        449.000000
75%        849.000000
max      39622.000000
Name: Launched Price(USA) in USD, dtype: float64

# Observation:-
#  Mean > Median → Strong right-skew, with a few high prices pulling up the average
# Interquartile Range (IQR): 599
# Suggests most phones are between 250–849 (budget to upper mid-range)

# Standard Deviation: 1,348.74
# Very high variation in prices, likely due to outliers

In [99]:
q1=df["Launched Price(USA) in USD"].quantile(0.25)
q3=df["Launched Price(USA) in USD"].quantile(0.75)
IQR=q3-q1
ll=q1-1.5*IQR
ul=q3+1.5*IQR
outliers=df["Launched Price(USA) in USD"][(df["Launched Price(USA) in USD"]<ll) | (df["Launched Price(USA) in USD"]>ul)]
print("Q1 = ",q1)
print("Q3 = ",q3)
print("IQR = ",IQR)
print("lower limit = ",ll)
print("upper limit = ",ul)
print("Outliers = ",outliers)
len(outliers)

Q1 =  250.0
Q3 =  849.0
IQR =  599.0
lower limit =  -648.5
upper limit =  1747.5
Outliers =  96      1799.0
115     1799.0
116     1899.0
119     1799.0
120     1899.0
630     2699.0
634     2499.0
643     2499.0
657     2499.0
660     2599.0
661     2799.0
682     1799.0
685    39622.0
708     1799.0
719     1799.0
836     1799.0
858     1899.0
869     1899.0
927     1899.0
929     2259.0
Name: Launched Price(USA) in USD, dtype: float64


20

# Observation:- Total 20 outlier in "Launched Price(USA) in USD"

In [100]:
df.head(1)

Company Name       Model Name  Mobile Weight(in gram)  RAM(in GB)  \
0        Apple  iPhone 16 128GB                   174.0         6.0   

  Front Camera Back Camera   Processor  Battery Capacity(in mAh) Screen Size  \
0         12MP        48MP  A17 Bionic                      3600  6.1 inches   

   Launched Price(Pakistan) in PKR  Launched Price(India) in INR  \
0                           224999                         79999   

   Launched Price(China) in CNY  Launched Price(USA) in USD  \
0                          5799                       799.0   

  Launched Price (Dubai)  Launched Year  
0              AED 2,799           2024

In [101]:
df["Launched Price (Dubai)"].unique().tolist()

['AED 2,799',
 'AED 2,999',
 'AED 3,199',
 'AED 3,399',
 'AED 3,599',
 'AED 3,499',
 'AED 3,699',
 'AED 3,899',
 'AED 3,799',
 'AED 3,999',
 'AED 4,199',
 'AED 2,699',
 'AED 4,399',
 'AED 4,499',
 'AED 2,599',
 'AED 3,299',
 'AED 4,599',
 'AED 2,499',
 'AED 4,999',
 'AED 4,299',
 'AED 4,799',
 'AED 5,099',
 'AED 3,099',
 'AED 2,099',
 'AED 1,199',
 'AED 1,499',
 'AED 1,799',
 'AED 5,299',
 'AED 5,799',
 'AED 6,099',
 'AED 4,899',
 'AED 6,999',
 'AED 7,499',
 'AED 4,099',
 'AED 6,499',
 'AED 7,099',
 'AED 1,399',
 'AED 1,299',
 'AED 1,099',
 'AED 899',
 'AED 999',
 'AED 1,699',
 'AED 799',
 'AED 699',
 'AED 499',
 'AED 5,499',
 'AED 1,999',
 'AED 599',
 'AED 2,199',
 'AED 1,899',
 'AED 399',
 'AED 2,899',
 'AED 2,299',
 'AED 2,399',
 'AED 4,699',
 'AED 1,649',
 'AED 549',
 'AED 649',
 'AED 1,599',
 'AED 749',
 'AED 5,199',
 'AED 1,000',
 'AED 1,200',
 'AED 1,300',
 'AED 1,100',
 'AED 950',
 'AED 900',
 'AED 850',
 'AED 800',
 'AED 750',
 'AED 700',
 'AED 1,500',
 'AED 1,700',
 'AED 1,40

In [102]:
df["Launched Price (Dubai)"].nunique()

141

# Observation:- Total 141 unique values are presemt in column "Launched Price (Dubai)"

In [103]:
df.rename(columns={"Launched Price (Dubai)":"Launched Price(Dubai) in AED"},inplace=True)

# Observation:- Renamed the column from "Launched Price (Dubai)" to "Launched Price(Dubai) in AED".

In [104]:
df["Launched Price(Dubai) in AED"]=df["Launched Price(Dubai) in AED"].str.replace("AED","")

# Observation:- Removed the value "AED" from column "Launched Price(Dubai) in AED".

In [105]:
df["Launched Price(Dubai) in AED"]=df["Launched Price(Dubai) in AED"].str.replace(",","")

# Observation:- Removed "," from column "Launched Price(Dubai) in AED"

In [106]:
df["Launched Price(Dubai) in AED"]=df["Launched Price(Dubai) in AED"].astype(int)

# Observation:- Convert the datatype from string to int.

In [107]:
df["Launched Price(Dubai) in AED"].describe()

count      928.000000
mean      2178.943966
std       1563.556403
min        299.000000
25%       1000.000000
50%       1675.000000
75%       3199.000000
max      11099.000000
Name: Launched Price(Dubai) in AED, dtype: float64

# Observation:-
# Since mean > median, the data is right-skewed (influenced by higher-priced models)
# Interquartile Range (IQR): AED 2,199, indicating wide price variation in mainstream devices

# Standard Deviation: AED 1,563.56
# Moderate to high variation, supporting the presence of diverse pricing tiers

In [108]:
q1=df["Launched Price(Dubai) in AED"].quantile(0.25)
q3=df["Launched Price(Dubai) in AED"].quantile(0.75)
IQR=q3-q1
ll=q1-1.5*IQR
ul=q3+1.5*IQR
outliers=df["Launched Price(Dubai) in AED"][(df["Launched Price(Dubai) in AED"]<ll) | (df["Launched Price(Dubai) in AED"]>ul)]
print("Q1 = ",q1)
print("Q3 = ",q3)
print("IQR = ",IQR)
print("lower limit = ",ll)
print("upper limit = ",ul)
print("Outliers = ",outliers)
len(outliers)

Q1 =  1000.0
Q3 =  3199.0
IQR =  2199.0
lower limit =  -2298.5
upper limit =  6497.5
Outliers =  115     6999
116     7499
119     6499
120     7099
630     9999
634     8999
643     9999
657     9999
660    10499
661    11099
708     6999
719     6999
836     6599
858     6999
869     6999
927     7199
928     7699
929     8699
Name: Launched Price(Dubai) in AED, dtype: int32


18

# Observation:- Total 18 outlier in "Launched Price(USA) in USD"

In [109]:
df["Launched Year"] = pd.to_datetime(df["Launched Year"], format="%Y")

# Observation :- Convert the column into date time format.

In [110]:
df["Launched Year"].unique().tolist()

[Timestamp('2024-01-01 00:00:00'),
 Timestamp('2023-01-01 00:00:00'),
 Timestamp('2022-01-01 00:00:00'),
 Timestamp('2021-01-01 00:00:00'),
 Timestamp('2020-01-01 00:00:00'),
 Timestamp('2019-01-01 00:00:00'),
 Timestamp('2017-01-01 00:00:00'),
 Timestamp('2018-01-01 00:00:00'),
 Timestamp('2016-01-01 00:00:00'),
 Timestamp('2014-01-01 00:00:00'),
 Timestamp('2025-01-01 00:00:00')]

In [111]:
df["Launched Year"].nunique()

11

# Observation:- Total 11 unique year in "Launched Year" column.

In [112]:
df["Launched Year"].mode()

0   2024-01-01
Name: Launched Year, dtype: datetime64[ns]

# Observation:- Mostly cell phone are launched in 2024 year

In [113]:
df["Launched Year"].value_counts()

Launched Year
2024-01-01    292
2023-01-01    182
2022-01-01    149
2021-01-01    109
2020-01-01    101
2019-01-01     46
2018-01-01     21
2025-01-01     12
2017-01-01      9
2016-01-01      5
2014-01-01      2
Name: count, dtype: int64

In [114]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 928 entries, 0 to 929
Data columns (total 15 columns):
 #   Column                           Non-Null Count  Dtype         
---  ------                           --------------  -----         
 0   Company Name                     928 non-null    object        
 1   Model Name                       928 non-null    object        
 2   Mobile Weight(in gram)           928 non-null    float64       
 3   RAM(in GB)                       928 non-null    float64       
 4   Front Camera                     928 non-null    object        
 5   Back Camera                      928 non-null    object        
 6   Processor                        928 non-null    object        
 7   Battery Capacity(in mAh)         928 non-null    int32         
 8   Screen Size                      928 non-null    object        
 9   Launched Price(Pakistan) in PKR  928 non-null    int32         
 10  Launched Price(India) in INR     928 non-null    int32         
 11

In [115]:
df.to_excel("Cleaned mobile data set.xlsx")

In [116]:
df

Company Name            Model Name  Mobile Weight(in gram)  RAM(in GB)  \
0          Apple       iPhone 16 128GB                   174.0         6.0   
1          Apple       iPhone 16 256GB                   174.0         6.0   
2          Apple       iPhone 16 512GB                   174.0         6.0   
3          Apple  iPhone 16 Plus 128GB                   203.0         6.0   
4          Apple  iPhone 16 Plus 256GB                   203.0         6.0   
..           ...                   ...                     ...         ...   
925         POCO          Pad 5G 128GB                   571.0         8.0   
926         POCO          Pad 5G 256GB                   571.0         8.0   
927      Samsung  Galaxy Z Fold6 256GB                   239.0        12.0   
928      Samsung  Galaxy Z Fold6 512GB                   239.0        12.0   
929      Samsung    Galaxy Z Fold6 1TB                   239.0        12.0   

        Front Camera Back Camera            Processor  \
0               12MP        48MP           A17 Bionic   
1               12MP        48MP           A17 Bionic   
2               12MP        48MP           A17 Bionic   
3               12MP        48MP           A17 Bionic   
4               12MP        48MP           A17 Bionic   
..               ...         ...                  ...   
925              8MP         8MP  Snapdragon 7s Gen 2   
926              8MP         8MP  Snapdragon 7s Gen 2   
927  10MP, 4MP (UDC)        50MP   Snapdragon 8 Gen 3   
928  10MP, 4MP (UDC)        50MP   Snapdragon 8 Gen 3   
929  10MP, 4MP (UDC)        50MP   Snapdragon 8 Gen 3   

     Battery Capacity(in mAh)  Screen Size  Launched Price(Pakistan) in PKR  \
0                        3600   6.1 inches                           224999   
1                        3600   6.1 inches                           234999   
2                        3600   6.1 inches                           244999   
3                        4200   6.7 inches                           249999   
4                        4200   6.7 inches                           259999   
..                        ...          ...                              ...   
925                     10000  12.1 inches                            66220   
926                     10000  12.1 inches                            71220   
927                      4400   7.6 inches                           604999   
928                      4400   7.6 inches                           544999   
929                      4400   7.6 inches                            84999   

     Launched Price(India) in INR  Launched Price(China) in CNY  \
0                           79999                          5799   
1                           84999                          6099   
2                           89999                          6499   
3                           89999                          6199   
4                           94999                          6499   
..                            ...                           ...   
925                         23999                          2099   
926                         25999                          2299   
927                        164999                         13999   
928                        176999                         15999   
929                        200999                         17999   

     Launched Price(USA) in USD  Launched Price(Dubai) in AED Launched Year  
0                         799.0                          2799    2024-01-01  
1                         849.0                          2999    2024-01-01  
2                         899.0                          3199    2024-01-01  
3                         899.0                          3199    2024-01-01  
4                         949.0                          3399    2024-01-01  
..                          ...                           ...           ...  
925                       280.0                          1029    2024-01-01  
926 

In [118]:
df['Screen Size'].value_counts()

Screen Size
6.7 inches               124
6.5 inches                76
6.67 inches               68
6.6 inches                64
6.78 inches               61
                        ... 
8.0 inches (unfolded)      1
7.8 inches (unfolded)      1
5.0 inches                 1
7.93 inches                1
6.79 inches                1
Name: count, Length: 97, dtype: int64

In [119]:
df['Screen Size'].unique()

array(['6.1 inches', '6.7 inches', '5.4 inches', '5.8 inches',
       '6.5 inches', '10.9 inches', '10.2 inches', '7.9 inches',
       '11 inches', '12.9 inches', '13 inches', '6.8 inches',
       '6.6 inches', '7.6 inches', '6.4 inches', '6.9 inches',
       '6.3 inches', '5.3 inches', '6.0 inches', '5.5 inches',
       '5.7 inches', '5.2 inches', '14.6 inches', '12.4 inches',
       '8.7 inches', '10.5 inches', '8 inches', '10.1 inches',
       '6.74 inches', '6.72 inches', '7.8 inches', '6.55 inches',
       '6.43 inches', '6.49 inches', '6.52 inches', '6.78 inches',
       '6.59 inches', '6.44 inches', '6.41 inches', '6.01 inches',
       '6.67 inches', '6.28 inches', '11.61 inches', '6.31 inches',
       '6.58 inches', '6.38 inches', '6.56 inches', '5.88 inches',
       '6.22 inches', '5.0 inches', '6.51 inches', '6.35 inches',
       '6.53 inches', '6.39 inches', '6.47 inches', '10.4 inches',
       '12.3 inches', '7.82 inches', '6.83 inches', '11.6 inches',
       '12.1 inches',